### ISSUE TITLE
Save Response Package PDF for release

### ISSUE THREAD

**Issue by lmullane** [bcgov/foi-flow#3231]
* As an IAO analyst
*I want to be able to accept all redactions
* So that I can save the response package (in the S3 bucket) so that the package can be released to the applicant.

**Assumptions & Scope**	
What are the assumptions for this story?

- Ministry Sign Off must be complete and the request in the Response state.
- Every page in the request package must have at least one of these flags applied: Partial Disclosure, Full Disclosure, Withheld in Full, Duplicate, Not Responsive 
- Packages with pages with no flag or the Consult, or In Progress and Left Off Here flag cannot be saved.
- All redactions will be applied in white.
- All records will be saved in one PDF document in Object Storage.
- The saved PDF will not include any pages that were flagged as Duplicate or Not Responsive
- The PDF document will include all redactions, applied in white, with the section stamp.
- The PDF document will include NO comments, highlights or annotations. 
- The PDF document will be viewable in the redaction app.
- All hyperlinks and embeds in the records package will be flattened/removed
- The Response package will be ordered based on file modified date. 

What is IN scope?
Saving the final records package with 

What is NOT in scope?
Creating redactions.

State changes.

Saved phased releases of the package will be in future story.

A dropdown to toggle between layers: 'Records Package', Redline, IAO Review, Harms, Ministry Sign Off

Creating a new records package.

**Acceptance Criteria**	

**Scenario 1: Inactive 'Final Package for Applicant" 1**
* GIVEN a records package is not in Response
* AND the ministry has not signed off
* AND every page in the Records Package has one of these page flags applied: Partial Disclosure, Full Disclosure, Withheld in Full, Duplicate, Not Responsive
* WHEN the IAO analyst clicks the "Create Records Package" dropdown
* THEN the dropdown will open with 2 options: "Redline for Sign Off" and "Final Package for Applicant"
https://jacklyn808742.invisionapp.com/console/share/8QVPHCNXSGA/981498459
* AND the "Redline for Sign Off" will be clickable
* AND the "Final Package for Applicant" will NOT be clickable

**Scenario 2: Active "Final Package for Applicant"**
* GIVEN a records package is in Response
* AND the ministry has signed off
* AND every page in the Records Package has one of these page flags applied: Partial Disclosure, Full Disclosure, Withheld in Full, Duplicate, Not Responsive
* WHEN the IAO analyst has clicked on the "Create Records Package" dropdown
* THEN the dropdown will open with 2 options: "Redline for Sign Off", "Final Package for Applicant"
https://jacklyn808742.invisionapp.com/console/share/8QVPHCNXSGA/981498459
* AND the "Redline for Sign Off" will be clickable
* AND the "Final Package for Applicant" will be clickable

**Scenario 3: Click "Final Package for Applicant"**
* GIVEN the Create Records Package" dropdown is open
* AND the request is in Response
* WHEN the IAO analyst clicks the active "Final Package for Applicant"
* THEN a modal will open with the message, "Are you sure want to create the records package for the applicant. This will apply all redactions. This should only be done when are redactions are finalized and are ready to send to the applicant. This will permanently apply the redactions and automatically create page stamps." 
* AND a "Create Applicant Package" and "Cancel" button
 https://jacklyn808742.invisionapp.com/console/share/8QVPHCNXSGA/981498459

**Scenario 4: Click "Cancel" button**
* GIVEN an analyst is on the "Create Package for Applicant" modal
* WHEN the IAO analyst clicks the "Cancel" button
* THEN a modal will close
* AND the analyst will remain in the redaction app
* AND no PDF will be saved

**Scenario 5: Click "Create Applicant Package" button**
* GIVEN an analyst is on the "Create Package for Applicant" modal
* WHEN the IAO analyst clicks the "Create Applicant Package" button
* THEN a modal will close
* AND a green success toast will appear when the save is complete
* AND the analyst will remain in the redaction app
* AND the analyst will see the created PDF in the redaction app so they can review if needed
* AND one PDF with all records will be saved in Object Storage
* AND the PDF will contain all redactions (a white box with the section of the ACT stamped in the box)
* AND the PDF will contain no comments or annotations of any kind
* AND the PDF will contain no pages that have been flagged as Not Responsive or Duplicate

**Dependencies? What is the impact of this dependency? (If so, link dependency in the ticket, make it visible in a team´s backlog)**

**Validation Rules? (If yes, list here)**	

**Design**
@xxx - please link the Design here

**Definition of Ready**

1. [ ] Is there a well articulated User Story?
2. [ ] Is there Acceptance Criteria that covers all scenarios (happy/sad paths)?
3. [ ] If there is a user interface, is there a design? 
4. [ ] Does the user story need user research/validation?
5. [ ] Does this User Story needs stakeholder approval?
6. [ ] Design / Solution accepted by Product Owner
7. [ ] Is this user story small enough to be completed in a Sprint? Should it be split?
8. [ ] Are the dependencies known/ understood? (technical, business, regulatory/policy)
9. [ ] Has the story been estimated?

**Definition of Done**

1.    [ ] Passes developer unit tests
2.    [ ] Passes peer code review
3.    [ ] If there's a user interface, passes UX assurance
4.    [ ] Passes QA of Acceptance Criteria with verification in Dev and Test
5.	[ ] Confirm Test cases built and succeeding 
6.	[ ] No regression test failures
7.	[ ] Test coverage acceptable by Product Owner
8.	[ ] Ticket ready to be merged to master or story branch
9.	[ ] Developer to list Config changes/ Update documents and designs
10.	[ ] Can be demoed in Sprint Review 
11.	[ ] Tagged as part of a Release
12.   [ ] Feature flagged if required
13.  [ ] Change Management activities done?

**Comment by lmullane** [comment]
@richard-aot, @m-prodan, @liseandtea, please review and revise accordingly.

cc @abin-aot, @KyEggleston

**Comment by m-prodan** [comment]
Moving back to product backlog, additional requirements likely to come - story will need further refinement and updates to AC

**Comment by m-prodan** [comment]
Moving back to product backlog, additional requirements likely to come - story will need further refinement and updates to AC

**Comment by lmullane** [comment]
@m-prodan, @richard-aot, @JHarrietha-AOT, @KyEggleston, @liseandtea 

I have updated the AC and moved this story back to new issues. The AC changes capture the dependency with page flags (#3413). Please review and advise if there is anything I missed.

@richard-aot, let us know if you have questions of if the AC updates change the estimate?

cc @abin-aot

**Comment by JHarrietha-AOT** [comment]
@lmullane My only thought it the actual style of the button, Would we want it to be more inline with the design of elements in PDF Tron so it looks like the "Left Off Here" Button on the top? Or do you want it to be inline with the FOI Mod Tracking System, which is how it is written in the story now?

My consern with having it coloured and blue in the redaction app is that nothing else is, and it will draw the users eye in for an action they only take at the end of the process

**Comment by lmullane** [comment]
Good catch, @JHarrietha-AOT. I'll revise to align with PDFTron. Is inactive Grey and Active Black?

**Comment by m-prodan** [comment]
I wonder if its a PDFTron function - but a page counter on the bottom for the response package is also a typical item we offer for applicants (it also helps for future review for page references). I don't know if it needs to be a scenario, or if its something out of the box? Maybe something we can talk with the devs tomorrow at refinement @lmullane

**Comment by m-prodan** [comment]
Note - the create records package will be a background job - we will need to create a story to address a notification to confirm the package is ready for release

**Comment by m-prodan** [comment]
Need to create a seperate user story for page numbering and stamping

**Comment by m-prodan** [comment]
Estimate should stay at 21 for this one

**Comment by abin-aot** [comment]
As indicated on refinement UX need to bit more elaborated with background running job idea for knitting the all documents and most importantly there is a need for story before this for "applying redactions on all Docs with a single button" and track that status on DB. Idea is to cover the validations and assumptions on this story. cc: @lmullane , @m-prodan , @JHarrietha-AOT .  Keeping same estimate for now
UX experience needs to be discussed. Because its a long running process, need to be completed as a BG running task and get alerted one done.Knitting, removing of pages can take time.

**Comment by lmullane** [comment]
@JHarrietha-AOT, just flagging that we need a design ticket for revisions to this story.

**Comment by lmullane** [comment]
Need to ensure that hyperlinks are not visible/recoverable. For example, if an email is redacted, the email hyperlink should no longer be clickable.

**Comment by JHarrietha-AOT** [comment]
@lmullane @m-prodan 

I have combined the designs for #3231 & #3437. I am hoping to avoid adding multiple buttons thats don't have a CTA that is used throughout the redaction process. It is something that is used to generate the files at specific times. I noticed this in the 4 tickets that have been these quick turn around designs - they were all to add additional buttons. When adding a button for each action isn't always the best solution, so to help ensure that we leave room for growth on the screen and also don't confuse users with way to main primary and secondary buttons like AXIS has. 

Let me know if you have any questions or comments:

ACTIVE: https://jacklyn808742.invisionapp.com/console/share/8QVPHCNXSGA/981498459

INACTIVE: https://jacklyn808742.invisionapp.com/console/share/8QVPHCNXSGA/981498460 
Because user has note assigned flags to every page

**Comment by lmullane** [comment]
Designs are great, @JHarrietha-AOT. 

I have updated the AC to reflect the designs. Assumption is that the analyst may need to create another Redline when the request is in Ministry Sign Off. @m-prodan is that correct?

**Comment by KyEggleston** [comment]
@JHarrietha-AOT - 'Create Package for Applicant' modal updates are below in bold and italics:

"This should only be done when all redactions are finalized and ready to **_be_** sent to the _**Applicant**._ This will **_permanently_** apply the redactions and automatically create page stamps."

**Comment by lmullane** [comment]
Updated the assumptions that a final package cannot be saved if any pages have an 'In Progress' flag.

**Comment by richard-aot** [comment]
Some AC's in Scenario 5 can only be QA after #3302.

**Comment by nimya-aot** [comment]
Will test this after @divyav-aot fix the issue with comments disappearing on redaction app

**Comment by KyEggleston** [comment]
Assumption that final records package will be ordered by file modified date - test as part of QA

**Comment by KyEggleston** [comment]
Note - withheld in full pages should not be displayed individually in the response package. Current AXIS process is one page showing a summary of all pages withheld in full, next page is xx.

**Comment by liseandtea** [comment]
All AC passes, however there is one minor issue which should be fixed by #4540. Sections are being cut off in the response package.
